[docs] Clarify SIL level TBAA rules.

atrick opened the pull request:
This was reviewed a while back by @jckarter. I'm just getting around to cleaning up the docs.

gottesmm commented:
I took a quick look. LGTM.

rudkx commented on docs/SIL.rst:
```diff
@@ -1538,20 +1538,23 @@ Class TBAA
+  non-class-instance heap objects.
+* Two values of the same class type ``$C`` may alias. Two values of
+  related class type ``$B`` and ``$D``, where there is a subclass
+  relationship between ``$B`` and ``$D``, may alias. Two values of of
```
typo: 'of of'

rudkx commented on docs/SIL.rst:
```diff
@@ -1538,20 +1538,23 @@ Class TBAA
+  relationship between ``$B`` and ``$D``, may alias. Two values of of
+  unrelated class types may not alias if typed access on both values
+  (via a ref_element_addr) are visible to the SIL program. This
+  includes different instantiations of a generic class type, such as
```
I'm not sure I know what you're getting at here. Can you clarify?

I might be reading it wrong, but as worded it sounds like I need to see the typed access for these to not alias, whereas if I don't see that (e.g. they are parameters), they do alias (perhaps because of casts through Builtin.{Native,Unknown,Bridge}Object.

atrick replied:
> On Apr 11, 2016, at 10:36 PM, Mark Lacey <notifications@github.com> wrote:
> 
> In docs/SIL.rst <https://github.com/apple/swift/pull/2147#discussion_r59323025>:
> 
> > -* A ``Builtin.UnknownObject`` may alias any class instance, whether Swift or
> > -  Objective-C, but may not alias non-class-instance heap objects.
> > -* Two values of the same class type ``$C`` may alias. Two values of related
> > -  class type ``$B`` and ``$D``, where there is a subclass relationship between
> > -  ``$B`` and ``$D``, may alias. Two values of unrelated class types may not
> > -  alias. This includes different instantiations of a generic class type, such
> > -  as ``$C<Int>`` and ``$C<Float>``, which currently may never alias.
> > +* A ``Builtin.UnknownObject`` or ``Builtin.BridgeObject`` may alias
> > +  any class instance, whether Swift or Objective-C, but may not alias
> > +  non-class-instance heap objects.
> > +* Two values of the same class type ``$C`` may alias. Two values of
> > +  related class type ``$B`` and ``$D``, where there is a subclass
> > +  relationship between ``$B`` and ``$D``, may alias. Two values of of
> > +  unrelated class types may not alias if typed access on both values
> > +  (via a ref_element_addr) are visible to the SIL program. This
> > +  includes different instantiations of a generic class type, such as
> I'm not sure I know what you're getting at here. Can you clarify?
> 
> I might be reading it wrong, but as worded it sounds like I need to see the typed access for these to not alias, whereas if I don't see that (e.g. they are parameters), they do alias (perhaps because of casts through Builtin.{Native,Unknown,Bridge}Object.
> 

I came across situations where the stdlib casts to class types that aren’t statically related. I wanted to make sure those cases are covered by the spec, because I don’t think we want to say that the stdlib has undefined behavior.

I removed the line above from the latest rewrite because I think it’s was conflating the two kinds of TBAA. I think the issue is better covered by the side note below (basically, the optimizer can pass around invalid class types as long as it doesn’t dereference them):

A violation of the above aliasing rules only results in undefined
behavior if the aliased types are actually accessed *statically*
within Swift code. For example,
``_SwiftNativeNS[Array|Dictionary|String]`` classes alias with
``NS[Array|Dictionary|String]`` classes even though they are not
statically related. Since Swift never directly accesses stored
properties on the Foundation classes, this aliasing does not pose a
danger.

Andy

rudkx commented on docs/SIL.rst:
```diff
@@ -1576,10 +1588,22 @@ the stored object or referent is not an allocated object of the relevant type.
+Although some raw pointer producing intrinsics exist, they do not
+affect TBAA. Builtin.inttoptr produces a RawPointer which is not
+interesting because by definition may alias with
+everything. Similarly, Builtin.bitcast and
```
s/may alias/it may alias/

?

atrick committed:
[docs] Clarify SIL-level TBAA rules.

This documentation now reflects the optimizer's reality.

I'm still working on formalizing language level rules for strict aliasing. Those
will be introduced in a separate doc file.

There some language about alias-introducing operations. This concept existed in
the old documentation but was never really implemented. This all makes sense now
that we have a formal model for binding memory to a type along with specific variants
of pointer_to_address that either enforce strict aliasing or permit type punning.

The detailed explanation of TBAA should probably be moved into a separate
optimizer document, but there isn't a good place for it yet.

atrick commented:
The previous version of this pull request was a little muddled (I had copied it in from an old document, and upon rereading I don't think it was clarifying much). I rewrote the language last night to reflect my current understanding. Hopefully it's clear now. Please review.

CodaFi commented:
@atrick Are these changes still relevant?

atrick commented:
This can *finally* be updated now that we have a formal notion of binding memory to a type. I just need a little more time to rewrite the language accordingly.

atrick commented:
I'm resurrecting this old documentation PR now that practice is catching up with theory.
